runtime/bundles/org.eclipse.e4.core.di/src/org/eclipse/e4/core/internal/di/InjectorImpl.java

```diff
@@ -425,7 +416,7 @@ else if (internalRequestor.isOptional())
 	}
 
 	public void disposed(PrimaryObjectSupplier objectSupplier) {
-		List<WeakReference<?>> references = getSupplierObjects(objectSupplier);
+		Set<IdentityWeakReference<?>> references = getSupplierObjects(objectSupplier);
```

Comment:
Below we are iterating the set of references outside of the synchronized block. It could be modified concurrently, if I’m understanding this correctly. This was already the case in the old impl. Now I’m wondering why we need the synchronization at all given that it’s flawed.

Reply:
Another thing: should we remove the refs that have been cleared already? Line 424 could use an iterator and call iterator.remove in the else branch of the nested if

Reply:
We're iterating over a set/list specific to one supplier though right?  So could another thread really be doing other things with that same supplier at the same time?  I can't say I understand the deeper behavior involved here, so maybe that is possible...

Reply:
Yes, I think @laeubi already mentioned that there were/are weak reference leaks.  But doesn't the final call to `forgetSupplier(objectSupplier)` just forget the whole list anyway?

Reply:
It’s certainly a huge improvement in its current shape. Nice work!

The overall design is hard to assess and I’ll leave remarks about the applied synchronization scheme aside.